### PR TITLE
@cambly/syntax-floating-components: Tooltip + Popover v2.0.0, built on top of AriaKit 

### PR DESCRIPTION
**Pull request by benjroy**
# Changes

This PR rewrites `Tooltip` and `Popover` on top of `Ariakit`.  This will introduce a breaking change to this repository

# Why?

We have bugs in our current Tooltip + Popover, as well as issues in mixing our components from this repo with similar components from material-ui.  @christianvuerings showed me [a new one](https://github.com/Cambly/Cambly-Frontend/pull/6556) just this morning

This PR is here to simplify the interfaces to these components + introduce something we can build on top of to save us a ton of development time.

------

The main breaking change is the shape of the components themselves.

Instead of three components that must be used together, e.g. `Tooltip  / TooltipTrigger / TooltipContent` there is only one:

```tsx
<Tooltip content="some tooltip content">
  <Link href="https://news.ycombinator.com" text="hover or focus me to see tooltip, click to use the link to navigate" />
</Tooltip>
```

In this PR, the thing I am most concerned about is the interfaces we ship to these components. We should make them as minimal and stable as possible.  That will allow us to continue making improvements or adjusting behavior internally in these comopnents without having a large impact to the callsites where they exist in our app.

Take the Tooltip + Popover for a spin [at the storybook build for this PR](https://vercel.live/open-feedback/syntax-git-20230907floatingcomponentsv2builtonariakit-cambly.vercel.app?via=pr-comment-visit-preview-link&passThrough=1)

## About AriaKit

Ariakit is an open-source project with deep development history whose goal is to make it easy to develop fully accessible applications.  It can hit two birds with one stone for us: 

1) reduce the time to develop the interactive components we have been discussing in our Syntax working group
2) ensure that all of the interactivity we build into our app will be accessible to our blind or differently-abled users.  This means screen readers and the standard tools that those users rely on.

The difference to user experience is immense and also the abstractions will allow us to write more immediately usable components with smaller prop interfaces.

### Browser accessibility quirks

`AriaKit` has _deep_ knowledge and handling of the quirks that exist in different environments.  For example, [here is some quirks handling for safari](https://github.com/ariakit/ariakit/blob/3ae4368de10f1b703bff6723ec33def4831887a8/packages/ariakit-react-core/src/dialog/dialog.tsx#L204-L231) re: tabbing & focus behavior in `Safari/VoiceOver`:

If we choose not to rely on a library like this, we will absolutely run into a subset of all the edge cases that AriaKit is already handling.  These quirks and issues will take us weeks of development time to research/understand/fix, and often through our research we will eventually find the fix in the ariakit repo itself.

### SSR

Ariakit (and Reakit before it) were designed from the start with ServerSideRendering in mind.  That's why in 2019 we chose to rely on it at my previous job.  The repo is not stale, and is still very active and responsive to issues and community PRs.

Here's [an example](https://ariakit.org/examples/dialog-next-router) describing an SSR flow with NextJS, where a dialog is rendered on the the server and immediately interactable on the client

### Bundle size comparisons:

We'll ship a few extra kilobytes compared to what we have right now, but the difference is small.
AriaKit builds on top of `floating-ui`, just like the current implementations here in `@cambly/syntax-floating-components`

[Tooltip v1 bundle: ](https://bundlejs.com/?q=%40cambly%2Fsyntax-floating-components&treeshake=%5B%7B+Tooltip+%7D%5D)
Bundle size is 185 kB -> 60.7 kB (gzip)

[Popover v1 bundle: ](https://bundlejs.com/?q=%40cambly%2Fsyntax-floating-components&treeshake=%5B%7B+Popover+%7D%5D)
Bundle size is 185 kB -> 60.7 kB (gzip)

[Tooltip v1 + Popover v1 bundle: ](https://bundlejs.com/?q=%40cambly%2Fsyntax-floating-components&treeshake=%5B%7B+Tooltip%2CPopover+%7D%5D)
Bundle size is 186 kB -> 60.7 kB (gzip)

[Popover components from AriaKit: ](https://bundlejs.com/?q=%40ariakit%2Freact&treeshake=%5B%7B%0A++usePopoverStore%2CPopover+as+AriakitPopover%2CPopoverDisclosure%2CPopoverArrow%2Ctype+PopoverStoreProps%2C%7D%5D)
Bundle size is 202 kB -> 66.9 kB (gzip)

[Tooltip Components from Ariakit: ](https://bundlejs.com/?q=%40ariakit%2Freact&treeshake=%5B%7B%0A++useTooltipStore%2CTooltip+as+AriakitTooltip%2CTooltipAnchor%2CTooltipArrow%2Ctype+TooltipStoreProps%2C%7D%5D)
Bundle size is 207 kB -> 68.4 kB (gzip)

[Tooltip + Popover components from aria kit:](https://bundlejs.com/?q=%40ariakit%2Freact%2C%40ariakit%2Freact&treeshake=%5B%7B%0A++usePopoverStore%2CPopover+as+AriakitPopover%2CPopoverDisclosure%2CPopoverArrow%2Ctype+PopoverStoreProps%2C%7D%5D%2C%5B%7B%0A++useTooltipStore%2CTooltip+as+AriakitTooltip%2CTooltipAnchor%2CTooltipArrow%2Ctype+TooltipStoreProps%2C%7D%5D)
Bundle size is 211 kB -> 69.5 kB (gzip)

The following is a little GPT output from phind.com on some quick selling points for the library:

- **Accessibility**: Ariakit is built with accessibility in mind, following WAI-ARIA recommendations. This ensures your interfaces will be accessible to all users, including those with disabilities [2][][3][][13][][14][].

- **Composability and Customizability**: Ariakit components can be easily composed to create new components, and they are fully customizable. This allows you to integrate them into any design or styling framework [3][][9][].

- **Adaptive**: Ariakit supports various input modes, making your UI components adaptable to different user needs and assistive technologies. It ensures consistent behavior across different UIs and has been tested across a variety of browsers, devices, and platforms [3][][9][].

- **React Hooks**: Ariakit provides a set of React Hooks designed to help developers build accessible UI components. These hooks include features like focus management and accessible labels [3][].

- **Convenience and Efficiency**: Ariakit handles many common accessibility considerations, allowing developers to focus more on the design and functionality of their applications. This can save time and effort in the development process [3][][10][].

- **Open and Community Support**: Ariakit is an open-project, benefiting from the contributions and support of a community of developers. This can lead to regular updates, improvements, and a wealth of resources and solutions for common issues [2][][14][].

Let's look at some code examples of using Ariakit:

**Creating an Accessible Button**

```jsx
import { Button } from "@ariakit/button";

function MyButton() {
  return <Button>Click me!</Button>;
}
```

**Creating an Accessible Dialog** **

```jsx
import "./style.css";
import * as Ariakit from "@ariakit/react";

export default function Example() {
  const dialog = Ariakit.useDialogStore();
  return (
    <>
      <Ariakit.Button onClick={dialog.show} className="button">
        Show modal
      </Ariakit.Button>
      <Ariakit.Dialog store={dialog} className="dialog">
        <Ariakit.DialogHeading className="heading">
          Success
        </Ariakit.DialogHeading>
        <p className="description">
          Your payment has been successfully processed. We have emailed your
          receipt.
        </p>
        <div>
          <Ariakit.DialogDismiss className="button">OK</Ariakit.DialogDismiss>
        </div>
      </Ariakit.Dialog>
    </>
  );
}

```

**Creating an Accessible Menu**

```jsx
import "./style.css";
import * as Ariakit from "@ariakit/react";

export default function Example() {
  const menu = Ariakit.useMenuStore();
  return (
    <>
      <Ariakit.MenuButton store={menu} className="button">
        Actions
        <Ariakit.MenuButtonArrow />
      </Ariakit.MenuButton>
      <Ariakit.Menu store={menu} gutter={8} className="menu">
        <Ariakit.MenuItem className="menu-item" onClick={() => alert("Edit")}>
          Edit
        </Ariakit.MenuItem>
        <Ariakit.MenuItem className="menu-item">Share</Ariakit.MenuItem>
        <Ariakit.MenuItem className="menu-item" disabled>
          Delete
        </Ariakit.MenuItem>
        <Ariakit.MenuSeparator className="separator" />
        <Ariakit.MenuItem className="menu-item">Report</Ariakit.MenuItem>
      </Ariakit.Menu>
    </>
  );
}
```

[2]: https://github.com/ariakit/ariakit
[3]: https://ariakit.org/examples
[9]: https://ariakit-ariakit.vercel.app/guide/getting-started
[10]: https://ariakit.org/tags/toolbar
[13]: https://reakit.io/
[14]: https://opencollective.com/ariakit

### CODE REVIEW

File: packages/syntax-floating-components/package.json [hunk 1]
```diff
@@ -21,6 +21,8 @@
     "test:watch": "vitest"
   },
   "dependencies": {
+    "@ariakit/react": "^0.3.0",
+    "@cambly/syntax-core": "workspace:^",
     "@cambly/syntax-design-tokens": "workspace:0.9.0",
     "@floating-ui/react": "^0.24.0"
```
Comment (christianvuerings) [inline packages/syntax-floating-components/package.json:27]: @benjroy can we make sure to remove the `@floating-ui/react` dependency now?

File: packages/syntax-floating-components/src/ariakit-utils/useForwardFocus.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  );
+
+  useEffect(() => {
+    if (
+      !anchorRef.current ||
+      !anchorElement ||
+      anchorElement === anchorRef.current
+    )
+      return;
+    // ariakit Tooltip uses html data attr to determine if mouse moves closes the tooltip or not
```
Comment (christianvuerings) [inline packages/syntax-floating-components/src/ariakit-utils/useForwardFocus.ts:49]: @benjroy this doesn't feel great... seems like we need to know about Ariakit internals to make this work. Let's definitely add tests around this.

File: packages/syntax-floating-components/src/index.tsx [hunk 3]
```diff
@@ -1,11 +1,4 @@
...
 
-export {
-  Popover,
-  PopoverContent,
-  PopoverTrigger,
-  Tooltip,
-  TooltipContent,
-  TooltipTrigger,
-};
+export { Popover, Tooltip };
```
Comment (christianvuerings) [inline packages/syntax-floating-components/src/index.tsx:4]: @benjroy way cleaner, thank you

File: packages/syntax-floating-components/src/Popover/Popover.stories.tsx [hunk 4]
```diff
@@ -1,13 +1,115 @@
...
+        backgroundColor="gray100"
+        display="flex"
+        direction="column"
+        justifyContent="center"
+        alignItems="center"
+      >
+        <Story />
+      </Box>
+    ),
+  ],
```
Comment (christianvuerings) [inline packages/syntax-floating-components/src/Popover/Popover.stories.tsx:112]: @benjroy let's make sure all the properties listed here are still available on the component - e.g. `onOpenChange` / `strategy`

**Review by christianvuerings — COMMENTED** [review]
@benjroy thanks for putting this PR together! I like the simplified API and improved accessibility features.

# Ariakit vs other libraries

Curious why you picked Ariakit vs something like [React Aria](https://react-spectrum.adobe.com/react-aria/) or [Radix UI](https://www.radix-ui.com/).

## Number of weekly downloads:
https://npmtrends.com/@ariakit/react-vs-@radix-ui/react-tooltip-vs-@react-aria/tooltip
image

## Maintenance

https://npmcompare.com/compare/@ariakit/react,@radix-ui/primitive,react-aria

|  | [Ariakit](https://ariakit.org/) | [React Aria](https://react-spectrum.adobe.com/react-aria/) | [Radix UI](https://www.radix-ui.com/)
| --- | --- | --- | --- |
| Owned & maintained by | [1 person - Diego Daz](https://github.com/ariakit/ariakit/blob/main/license) | [Adobe](https://github.com/adobe/react-spectrum/blob/main/LICENSE) | [WorkOS](https://github.com/radix-ui/themes/blob/main/LICENSE)

What is going to happen to Ariakit if Diego no longer works on it?

## Other comparisons
* Side-by-side: https://npmcompare.com/compare/@ariakit/react,@radix-ui/primitive,react-aria
* Side-by-side: https://moiva.io/?npm=@ariakit/react+@radix-ui/primitive+react-aria
* Security: [@radix-ui/primitive](https://snyk.io/advisor/npm-package/@radix-ui/primitive) / [react-aria](https://snyk.io/advisor/npm-package/react-aria) / [@ariakit/react](https://snyk.io/advisor/npm-package/@ariakit/react) (all 90+)


# Todo's for this PR if we want to go ahead with Ariakit

- [ ]  Add SSR test files for both Popover and Tooltip - to avoid issues like we had with Layer: https://github.com/Cambly/syntax/pull/255
- [ ] Address comments in this PR
- [ ] Figure out root cause as to why hovering / focussing doesn’t keep a Popover or Tooltip open in Storybook 

https://github.com/Cambly/syntax/assets/127199/3607aac8-706e-434e-883e-920269a26183

# Action items after this PR lands:

- [ ] Update all Popover / Tooltip components in Cambly-Frontend

**Comment by benjroy** [comment]
@christianvuerings I haven't checked out radix yet, but from these results alone in a two way matchup b/w react-aria and ariakit, `react-aria` is the clear choice. 

Bundle size is 33.4 kB -> 11.2 kB (gzip):
[react-aria tooltip dependencies](https://bundlejs.com/?q=react-stately%2Creact-aria&treeshake=%5B%7BuseTooltipTriggerState%7D%5D%2C%5B%7BmergeProps%2CuseTooltip%2CuseTooltipTrigger%7D%5D) (used in an examples from their docs)

Bundle size is 154 kB -> 49.7 kB (gzip)
[ariakit tooltip dependencies](https://bundlejs.com/?q=ariakit&treeshake=%5B%7BTooltip%2CTooltipAnchor%2CTooltipArrow%7D%5D) (used in an example from their docs)

**Comment by diegohaz** [comment]
Hi! I happened upon this PR during my weekly search. As this project is open source, I want to mention that if you end up using Ariakit, please feel free to mention me on PRs/issues for reviews or any questions you might have.

A few notes:

- The entire Ariakit library is approximately [50 kB gzip](https://bundlejs.com/?q=%40ariakit%2Freact&treeshake=%5B*+as+Ariakit%5D&config=%7B%22esbuild%22%3A%7B%22external%22%3A%5B%22react%22%2C%22react-dom%22%5D%7D%7D), excluding the `react` and `react-dom` peer dependencies. The Tooltip is roughly 23 kB gzip, mainly because it incorporates numerous accessibility enhancements and addresses many edge cases. In fact, it's powered by [Hovercard](https://ariakit.org/components/hovercard) and [Dialog](https://ariakit.org/components/dialog). Tooltip shares a good amount of code with other modules in the library, so the increase in bundle size should be offset as the project uses other modules.

- Regarding maintenance, I suggest this read: [Ariakit is antifragile](https://newsletter.ariakit.org/p/stable#%C2%A7ariakit-is-antifragile)

- Here's an example of building the Radix Dialog API using Ariakit underneath: https://ariakit.org/examples/dialog-radix. The document delves deeper into some of the edge cases that Ariakit resolves.

File: packages/syntax-floating-components/package.json [hunk 5]
```diff
@@ -21,6 +21,8 @@
     "test:watch": "vitest"
   },
   "dependencies": {
+    "@ariakit/react": "^0.3.0",
```
Comment (christianvuerings) [inline packages/syntax-floating-components/package.json:24]: @benjroy could we make sure that we use the latest version of @ariakit/react v0.3.3

File: packages/syntax-floating-components/src/Popover/Popover.tsx [hunk 6]
```diff
@@ -1,263 +1,164 @@
...
+  type PopoverStoreProps,
+} from "@ariakit/react";
+import useForwardFocus from "../ariakit-utils/useForwardFocus";
+import useChangeContentVisibility from "../ariakit-utils/useChangeContentVisibility";
+
+type PopoverProps = {
+  /** Optional aria-label for the popover (content element) */
+  accessibilityLabel?: string;
+  /** Required trigger element */
+  children?: ReactElement | string;
```
[the remark this answers is not in the document]
Comment (benjroy) [reply packages/syntax-floating-components/src/Popover/Popover.tsx:23]: @christianvuerings I'm prepping the PR for these components on top of react-aria at the moment.  Going throught your old review here in this PR to see what comments still apply to that branch.  This one does.  When I change it to only `ReactElement` , it's too restrictive.  TS errors look like this:

image

File: packages/syntax-floating-components/src/Popover/Popover.tsx [hunk 7]
```diff
@@ -1,263 +1,164 @@
...
-
-const usePopoverContext = () => {
-  const context = React.useContext(PopoverContext);
-
-  if (context == null) {
-    throw new Error("Popover components must be wrapped in <Popover />");
-  }
-
-  return context;
+  placement?: PopoverStoreProps["placement"];
```
[the remark this answers is not in the document]
Comment (benjroy) [reply packages/syntax-floating-components/src/Popover/Popover.tsx:41]: @christianvuerings I think this will still be something to comment on in the new react-aria based pr.  We'll probably want to consolidate the Placement prop between Tooltip + Popover.  Keep an eye out for the discrepency in that PR and lmk if you have a lightbulb go off for a clean way to consolidate them 🙏

File: packages/syntax-floating-components/src/Popover/Popover.tsx [hunk 8]
```diff
@@ -1,263 +1,164 @@
...
+        preserveTabOrder
+        tabIndex={0}
+
+        // ariakit sets role
+      >
+        <>
+          <Box
+            position="relative"
+            width="100%"
+            maxHeight="var(--popover-available-height, 100%)"
```
[the remark this answers is not in the document]
Comment (benjroy) [reply packages/syntax-floating-components/src/Popover/Popover.tsx:137]: Aside:  That ariakit repo is set up quite well.  Things are easy to find and make sense once they've been read

File: packages/syntax-floating-components/src/Popover/Popover.tsx [hunk 9]
```diff
@@ -1,263 +1,164 @@
...
-   *
+   * If set to true the popover will render initially open
    * @defaultValue false
    */
   initialOpen?: boolean;
-  /**
-   * Value of 'open' state of the Popover content if controlled
-   */
+  /** Optional boolean to control whether popover content is rendered as a modal */
+  modal?: boolean;
```
[the remark this answers is not in the document]
Comment (benjroy) [reply packages/syntax-floating-components/src/Popover/Popover.tsx:32]: I think that this will be a comment you'll be making in the new PR.
